design/cluster-api-provider-metal3/allow_disabling_node_disk_cleaning.md

```diff
@@ -0,0 +1,234 @@
+<!--
+This work is licensed under a Creative Commons Attribution 3.0
+Unported License.
+
+http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# allow disabling disk cleaning
+
+## Status
+
+implementable
+
+## Summary
+
+Configurable disk cleaning interface in Metal³ to allow users to
+disable/enable disk cleaning for the nodes.
+
+## Motivation
+
+Currently, when provisioning & de-provisioning Ironic nodes go
+through the [automated-cleaning](https://docs.openstack.org/ironic/latest/admin/cleaning.html#automated-cleaning)
+operation by default. This operation wipes out all the available
+disks on every node. While upgrade operations in Kubernetes,
+we want to ensure we don’t lose the data on the disks attached to
+the nodes. As such, we need a way to disable automated cleaning per
+node so that we can avoid disk cleaning while upgrading the cluster node(s).
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I would like my node’s disks data kept untouched
+while I'm upgrading my Kubernetes cluster node(s).
+
+#### Story 2
+
+As a cluster admin, I would like my node’s disks to be cleaned
+while provisioning/de-provisioning (i.e. not upgrade scenario)
+operations.
+
+### Goals
+
+- Add a filtering mechanism in Cluster-api-provider-metal3 (CAPM3) to find out Machines
+   that should/not experience disk cleaning
+
+- To Baremetal-operator (BMO), instruct whether a BareMetalHost should have its disk
+   cleaned or not
+
+- To Ironic, instruct whether an Ironic node should have its disk cleaned or not
+
+### Non-Goals
+
+During the upgrade, a mechanism to select the exact same nodes after de-provisioning.
+In other words, nodes re-use logic.
+
+## Proposal
+
+Ironic supports enabling & disabling automated cleaning globally
+and on the node level. As such, Metal³ needs to be able to identify the nodes
+for which the disk cleaning should be disabled and communicate to Ironic to request
+disabling the disk cleaning for those specific nodes.
+
+### Metal³
+
+In the Metal³, we will need changes both in CAPM3 and BMO as follows
+
+#### Cluster-api-provider-metal3
+
+##### Metal3MachineTemplate object
+
+We introduce a new field `disableAutomatedClean` in the spec of the
+Metal3MachineTemplate object. `disableAutomatedClean` can store a boolean type
+of value. Setting it to True means don’t perform automated cleaning, while
+False value enables automated cleaning. By default, `disableAutomatedClean` will
+be set to False (i.e. do perform automated cleaning as it is now).
```

Comment:
We were going to use a pointer field, right? So the default will be nil, which is treated the same as false in this case.

Reply:
Yes, that should work perfectly fine.